examples/nft-nd-ns/ns.go

```diff
@@ -0,0 +1,655 @@
+	if nsBytes == nil {
+		return true
+	}
+	ns := std.Deserialize(nsBytes.([]byte)).(NameState)
+	return runtime.GetTime() >= ns.Expiration
+}
+
+// Register registers new domain with the specified owner and name if it's available.
+func Register(name string, owner interop.Hash160) bool {
+	fragments := splitAndCheck(name, false)
+	if fragments == nil {
+		panic("invalid domain name format")
+	}
+	ctx := storage.GetContext()
+	if storage.Get(ctx, append([]byte{prefixRoot}, []byte(fragments[1])...)) == nil {
+		panic("root not found")
+	}
+
+	if !isValid(owner) {
+		panic("invalid owner")
+	}
+	if !runtime.CheckWitness(owner) {
+		panic("not witnessed by owner")
+	}
+	runtime.BurnGas(GetPrice())
+	var (
+		tokenKey = getTokenKey([]byte(name))
+		oldOwner interop.Hash160
+	)
+	nsBytes := storage.Get(ctx, append([]byte{prefixName}, tokenKey...))
+	if nsBytes != nil {
+		ns := std.Deserialize(nsBytes.([]byte)).(NameState)
+		if runtime.GetTime() < ns.Expiration {
+			return false
+		}
+		oldOwner = ns.Owner
+		updateBalance(ctx, []byte(name), oldOwner, -1)
+	} else {
+		updateTotalSupply(ctx, +1)
+	}
+	ns := NameState{
+		Owner:      owner,
+		Name:       name,
+		Expiration: runtime.GetTime() + millisecondsInYear,
+	}
+	putNameStateWithKey(ctx, tokenKey, ns)
+	updateBalance(ctx, []byte(name), owner, +1)
+	postTransfer(oldOwner, owner, []byte(name), nil)
+	return true
+}
+
+// Renew increases domain expiration date.
+func Renew(name string) int {
+	if len(name) > maxDomainNameLength {
+		panic("invalid domain name format")
+	}
+	runtime.BurnGas(GetPrice())
+	ctx := storage.GetContext()
+	ns := getNameState(ctx, []byte(name))
+	ns.Expiration += millisecondsInYear
+	putNameState(ctx, ns)
+	return ns.Expiration
+}
+
+// SetAdmin updates domain admin.
+func SetAdmin(name string, admin interop.Hash160) {
+	if len(name) > maxDomainNameLength {
+		panic("invalid domain name format")
+	}
+	if admin != nil && !runtime.CheckWitness(admin) {
+		panic("not witnessed by admin")
+	}
+	ctx := storage.GetContext()
+	ns := getNameState(ctx, []byte(name))
+	if !runtime.CheckWitness(ns.Owner) {
+		panic("not witnessed by owner")
+	}
+	ns.Admin = admin
+	putNameState(ctx, ns)
+}
+
+// SetRecord adds new record of the specified type to the provided domain.
+func SetRecord(name string, typ RecordType, data string) {
+	tokenID := []byte(tokenIDFromName(name))
+	var ok bool
+	switch typ {
+	case A:
+		ok = checkIPv4(data)
+	case CNAME:
+		ok = splitAndCheck(data, true) != nil
+	case TXT:
+		ok = len(data) <= maxTXTRecordLength
+	case AAAA:
+		ok = checkIPv6(data)
+	default:
+		panic("unsupported record type")
+	}
+	if !ok {
+		panic("invalid record data")
+	}
+	ctx := storage.GetContext()
+	ns := getNameState(ctx, tokenID)
+	ns.checkAdmin()
+	putRecord(ctx, tokenID, name, typ, data)
+}
+
+// GetRecord returns domain record of the specified type if it exists or an empty
+// string if not.
+// TODO: C# node returns stackitem.Null{} in case of an empty record, but
```

Comment:
What will our compiler generate for `return string([]byte(nil))`?

Reply:
Null, so it suits our purposes, I'll fix that.

Reply:
Fixed.